File: Linked List/21. Merge Two Sorted Lists/mergeTwoLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// list1 = 1 -> 2 -> 4
// list2 = 1 -> 3 -> 4
// ans = (-1)
// ans = -1 -> 1 (l1) if l1->val <= l2->val
//     = -1 -> 1 -> 1 (l2) if l1->val > l2->val
//       -1 -> 1 -> 1 -> 2, if l1->val <= l2->val (2 < 3)
//       -1 -> 1 -> 1 -> 2 -> 3 -> 4 -> 4 if(l1 == NULL) ans->next = l2;       

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* ans = new ListNode (-1);
        ListNode* pilot = ans;
        while(list1 != NULL && list2 != NULL) {
            if(list1->val <= list2->val) {
                pilot->next = list1;
                pilot = pilot->next;
                list1 = list1->next;
            } else {
                pilot->next = list2;
                pilot = pilot->next;
                list2 = list2->next;
            }
        }
        pilot->next = list2 == NULL ? list1 : list2;
        return ans->next;
    }
};

File: Linked List/21. Merge Two Sorted Lists/mergeTwoLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
'''
 list1 = 1 -> 2 -> 4
 list2 = 1 -> 3 -> 4
 ans = (-1)
 ans = -1 -> 1 (l1) if l1->val <= l2->val
     = -1 -> 1 -> 1 (l2) if l1->val > l2->val
       -1 -> 1 -> 1 -> 2, if l1->val <= l2->val (2 < 3)
       -1 -> 1 -> 1 -> 2 -> 3 -> 4 -> 4 if(l1 == NULL) ans->next = l2;    
'''
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # list1 = [1,2,4], list2 = [1,3,4]
        if list1 is None:
            return list2
        elif list2 is None:
            return list1
        elif list1.val <= list2.val:
            # 1. mergeTwoLists([2,4], [1,3,4]); 3. mergeTwoLists([4], [3,4]); 5. mergeTwoLists([], [4])
            list1.next = self.mergeTwoLists(list1.next, list2)
            # list1.next = list2 [4]; list1.next = list2
            return list1 # list1 -> 4; list1 -> [3,4,4]
        else:
            # 2. mergeTwoLists([2,4], [3,4]); 4. mergeTwoLists([4], [4]); 
            list2.next = self.mergeTwoLists(list1, list2.next)
            # list2.next = list1 [4,4]
            return list2 # list2 -> [4,4]